File: aether_flow_apple/AetherNetwork.swift
import Foundation
import Network

/// AetherFlow 網路通訊層 (Swift/Apple)
/// 負責與已配對設備進行 TCP 連線，並透過 AetherCrypto 加密傳輸 Universal Clipboard 等資料
public class AetherNetwork: ObservableObject {
    private var listener: NWListener?
    private var connection: NWConnection?
    
    private let crypto: AetherCrypto
    private let clipboardMonitor: ClipboardMonitor
    
    // 用於 Demo 的固定 Port
    private let port: NWEndpoint.Port = 8080
    
    public init(cryptoKey: [UInt8], clipboardMonitor: ClipboardMonitor) throws {
        self.crypto = try AetherCrypto(keyBytes: cryptoKey)
        self.clipboardMonitor = clipboardMonitor
    }
    
    /// 啟動監聽服務 (作為接收端)
    public func startServer() {
        do {
            listener = try NWListener(using: .tcp, on: port)
            listener?.stateUpdateHandler = { state in
                print("🌐 [Network] Listener state: \(state)")
            }
            
            listener?.newConnectionHandler = { [weak self] newConnection in
                print("🌐 [Network] Received new connection from \(String(describing: newConnection.endpoint))")
                self?.connection = newConnection
                self?.startReceiving(on: newConnection)
                newConnection.start(queue: .main)
            }
            
            listener?.start(queue: .main)
            print("🌐 [Network] Server started on port \(port)")
        } catch {
            print("🌐 [Network] Failed to start server: \(error)")
        }
    }
    
    /// 建立連線到遠端設備 (透過 mDNS 解析出 IP 後呼叫)
    public func connectToRemote(host: String) {
        let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(host), port: port)
        connection = NWConnection(to: endpoint, using: .tcp)
        
        connection?.stateUpdateHandler = { state in
            print("🌐 [Network] Connection state: \(state)")
        }
        
        startReceiving(on: connection!)
        connection?.start(queue: .main)
    }
    
    /// 開始接收遠端傳來的資料
    private func startReceiving(on connection: NWConnection) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] content, context, isComplete, error in
            guard let self = self else { return }
            
            if let data = content {
                self.handleIncomingData(data)
            }
            
            if isComplete {
                print("🌐 [Network] Connection closed")
                return
            }
            
            if let error = error {
                print("🌐 [Network] Receive error: \(error)")
                return
            }
            
            // 繼續監聽
            self.startReceiving(on: connection)
        }
    }
    
    /// 處理收到的加密資料
    private func handleIncomingData(_ data: Data) {
        do {
            let encryptedBytes = Array(data)
            let decryptedBytes = try crypto.decryptPayload(encryptedBytes)
            
            if let stringPayload = String(bytes: decryptedBytes, encoding: .utf8) {
                print("🔒 [Crypto] Decrypted clipboard payload successfully")
                // 同步至本機剪貼簿
                clipboardMonitor.syncFromRemote(text: stringPayload)
            }
        } catch {
            print("🔒 [Crypto] Decryption failed: \(error)")
        }
    }
    
    /// 發送本機剪貼簿資料至遠端
    public func sendClipboardData(_ text: String) {
        guard let connection = connection, connection.state == .ready else {
            print("🌐 [Network] No active connection to send data.")
            return
        }
        
        guard let textBytes = text.data(using: .utf8).map({ Array($0) }) else { return }
        
        do {
            let encryptedBytes = try crypto.encryptPayload(textBytes)
            let data = Data(encryptedBytes)
            
            connection.send(content: data, completion: .contentProcessed { error in
                if let error = error {
                    print("🌐 [Network] Send error: \(error)")
                } else {
                    print("🔒 [Crypto] Encrypted and sent clipboard payload")
                }
            })
        } catch {
            print("🔒 [Crypto] Encryption failed: \(error)")
        }
    }
}

File: aether_flow_flutter/lib/aether_network.dart
import 'dart:io';
import 'dart:typed_data';
import 'aether_crypto.dart';
import 'clipboard_monitor.dart';

/// AetherFlow 網路通訊層 (Flutter/Dart)
/// 負責與已配對設備進行 TCP 連線，並透過 AetherCrypto 加密傳輸 Universal Clipboard 等資料
class AetherNetwork {
  ServerSocket? _serverSocket;
  Socket? _connection;
  
  final AetherCrypto _crypto;
  final ClipboardMonitor _clipboardMonitor;
  
  // 用於 Demo 的固定 Port
  final int port = 8080;

  AetherNetwork(this._crypto, this._clipboardMonitor);

  /// 啟動監聽服務 (作為接收端)
  Future<void> startServer() async {
    try {
      _serverSocket = await ServerSocket.bind(InternetAddress.anyIPv4, port);
      print("🌐 [Network] Server started on port $port");

      _serverSocket?.listen((Socket socket) {
        print("🌐 [Network] Received new connection from ${socket.remoteAddress.address}");
        _connection = socket;
        _startReceiving(socket);
      });
    } catch (e) {
      print("🌐 [Network] Failed to start server: $e");
    }
  }

  /// 建立連線到遠端設備
  Future<void> connectToRemote(String host) async {
    try {
      _connection = await Socket.connect(host, port);
      print("🌐 [Network] Connected to remote $host:$port");
      _startReceiving(_connection!);
    } catch (e) {
      print("🌐 [Network] Failed to connect: $e");
    }
  }

  /// 開始接收遠端傳來的資料
  void _startReceiving(Socket socket) {
    socket.listen(
      (Uint8List data) async {
        await _handleIncomingData(data);
      },
      onError: (error) {
        print("🌐 [Network] Receive error: $error");
        socket.destroy();
      },
      onDone: () {
        print("🌐 [Network] Connection closed");
        socket.destroy();
      },
    );
  }

  /// 處理收到的加密資料
  Future<void> _handleIncomingData(Uint8List data) async {
    try {
      final decryptedBytes = await _crypto.decryptPayload(data);
      final stringPayload = String.fromCharCodes(decryptedBytes);
      
      print("🔒 [Crypto] Decrypted clipboard payload successfully");
      // 同步至本機剪貼簿
      await _clipboardMonitor.syncFromRemote(stringPayload);
    } catch (e) {
      print("🔒 [Crypto] Decryption failed: $e");
    }
  }

  /// 發送本機剪貼簿資料至遠端
  Future<void> sendClipboardData(String text) async {
    if (_connection == null) {
      print("🌐 [Network] No active connection to send data.");
      return;
    }

    try {
      final textBytes = text.codeUnits;
      final encryptedBytes = await _crypto.encryptPayload(textBytes);
      
      _connection?.add(encryptedBytes);
      print("🔒 [Crypto] Encrypted and sent clipboard payload");
    } catch (e) {
      print("🔒 [Crypto] Encryption failed: $e");
    }
  }

  void dispose() {
    _connection?.destroy();
    _serverSocket?.close();
  }
}
